File: python/rotational-cipher/rotational_cipher.py
ALPA_NUM = 26


def rotate_char(code, code_a, key):
    return chr((code - code_a + key) % ALPA_NUM + code_a)


def rotate(text, key):
    result = ''
    char_a = ord('a')
    char_A = ord('A')
    for char in [*text]:
        if char.isalpha():
            ascii_code = ord(char)
            if char.isupper():
                result += rotate_char(ascii_code, char_A, key)
            else:
                result += rotate_char(ascii_code, char_a, key)
        else:
            result += char

    return result

File: c/rotational-cipher/rotational_cipher.c
#include "rotational_cipher.h"
#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define ALPHA_SIZE 26

char *rotate(const char *text, int shift_key){ 
    int length = strlen(text);
    char *output = calloc(length + 1, sizeof(char));
    if (!output) return NULL;

    shift_key = shift_key % ALPHA_SIZE;
    
    int index = 0;
    while(*text){
        char ch = *text++;
        if (isalpha(ch)){
            int ground = isupper(ch) ? 'A' : 'a';    
            ch = (ch + shift_key - ground) % ALPHA_SIZE + ground;
        }
        output[index++] = ch;
    }

    return output;
}
